### PR TITLE
NTRN-383 revision number and height in query

### PR DESCRIPTION
**Pull request by quasisamurai**
[TASK](https://p2pvalidator.atlassian.net/browse/NTRN-383)

[TEST](https://github.com/neutron-org/neutron-tests/actions/runs/4437847594)

There are bunch of PRs:

1.  [Neutron node](https://github.com/neutron-org/neutron/pull/165)
2. [Query relayer](https://github.com/neutron-org/neutron-query-relayer/pull/76)
3. [Integration tests](https://github.com/neutron-org/neutron-integration-tests/pull/102)
4. [Dev contracts](https://github.com/neutron-org/neutron-dev-contracts/pull/10)
5. [SDK](https://github.com/neutron-org/neutron-sdk/pull/85)

### CODE REVIEW

File: x/interchainqueries/keeper/keeper.go [hunk 1]
```diff
@@ -86,6 +86,14 @@ func (k Keeper) SetLastRegisteredQueryKey(ctx sdk.Context, id uint64) {
 func (k Keeper) SaveQuery(ctx sdk.Context, query *types.RegisteredQuery) error {
 	store := ctx.KVStore(k.storeKey)
 
+	// We're doing this because LastSubmittedResultRemoteHeight is a struct pointer type generated from proto
+	// to avoid nil in neutron, null in json and rust marshalling errors, here we initialize it with "default" values
+	// which is not equal 0 because github.com/cosmos/cosmos-sdk/codec/types skips true default values
+	emptyHeight := ibcclienttypes.NewHeight(1, 1)
+	if query.LastSubmittedResultRemoteHeight == nil {
```
Comment (mmikev) [inline x/interchainqueries/keeper/keeper.go]: Why not just to use `Option<u64>` on sc level?

File: x/interchainqueries/keeper/keeper.go [hunk 2]
```diff
@@ -244,17 +252,21 @@ func (k Keeper) updateLastLocalHeight(ctx sdk.Context, query *types.RegisteredQu
...
-// of the last result submission.
-func (k Keeper) checkLastRemoteHeight(ctx sdk.Context, query types.RegisteredQuery, height uint64) error {
-	if query.LastSubmittedResultRemoteHeight >= height {
+// of the last result submission. Also it checks if revision number has updated. If yes, we don't care about revision height
+// see
+func (k Keeper) checkLastRemoteHeight(ctx sdk.Context, query types.RegisteredQuery, height ibcclienttypes.Height) error {
+	if query.LastSubmittedResultRemoteHeight.RevisionNumber < height.GetRevisionNumber() {
+		return nil
+	} else if query.LastSubmittedResultRemoteHeight.RevisionHeight >= height.GetRevisionHeight() &&
+		query.LastSubmittedResultRemoteHeight.RevisionNumber == height.GetRevisionNumber() {
```
Comment (mmikev) [inline x/interchainqueries/keeper/keeper.go]: `ibcclienttypes.Height` has all the necessary methods to compare two structs (`LTE`, `GTE`, etc).

**Review by pr0n00gler — CHANGES_REQUESTED** [review]

File: x/interchainqueries/keeper/keeper.go (hunk 1 above)
Comment (quasisamurai) [reply x/interchainqueries/keeper/keeper.go]: if you meant `Option<Height>`,  I believe couple of simple lines in neutron node is a bit preferable than unnecessary complications in our smart contracts, also we should init this struct anyway (in other place), even w `Option`

**Comment by quasisamurai** [comment]
Talked to @swelf19, going too try smth to make it look a bit better

**Comment by quasisamurai** [comment]
[FRESH TEST RUN](https://github.com/neutron-org/neutron-tests/actions/runs/4480553574/jobs/7876250946)

**Review by swelf19 — DISMISSED** [review]

**Review by pr0n00gler — DISMISSED** [review]

**Review by pr0n00gler — APPROVED** [review]

**Review by swelf19 — APPROVED** [review]

**Pull request merged**